inc/redirect/namespace.php

```diff
@@ -217,15 +221,17 @@ function do_open() {
 
 	$action = get_query_var( 'open' );
 
-	if ( 'download' === $action ) {
+	if ( 'download' === $action && ! empty( $_GET['type'] ) ) {
 		// Download
-		if ( ! empty( $_GET['type'] ) ) {
-			$files = \Pressbooks\Utility\latest_exports();
-			if ( isset( $files[ $_GET['type'] ] ) ) {
-				$filepath = \Pressbooks\Modules\Export\Export::getExportFolder() . $files[ $_GET['type'] ];
-				force_download( $filepath );
-				exit;
-			}
+		$format = $_GET['type'];
+		$files = \Pressbooks\Utility\latest_exports();
+
+		if ( isset( $files[ $format ] ) ) {
+			do_action( 'store_download_data', $format );
+
+			$filepath = \Pressbooks\Modules\Export\Export::getExportFolder() . $files[ $format ];
+			$do_download( $filepath );
```

Comment:
Cara, should we ensure the callback function is not null?
`if ( ! is_null( $do_download ) )`, just in case we need to use this function and we don't send any callback function

Reply:
Yes, we could add something like this ` $do_download && $do_download( $filepath );` maybe?

Reply:
yes, makes sense to me!